File: python/tests/index_aggregates_test.py
import pytest  # https://docs.pytest.org/en/7.4.x/
from elasticsearch import Elasticsearch  # https://elasticsearch-py.readthedocs.io/

from ..app import elasticsearch_client

target_index = "test_index_2"


@pytest.fixture
def client():
    client = elasticsearch_client.get_client()
    return client


@pytest.mark.asyncio
async def index_document_count_test(client: Elasticsearch):
    index_names = ["test_index_1", "test_index_2"]
    query_size = 0  # Set the result set size to 0 to just get the aggregation results.
    aggs = {"indexes": {"terms": {"field": "_index", "size": len(index_names)}}}

    response = await client.search(
        index=",".join(index_names), aggs=aggs, size=query_size
    )
    aggregations = response.body["aggregations"]

    indexes = aggregations["indexes"]
    buckets = indexes["buckets"]
    assert len(buckets) == len(index_names)

    first_bucket = buckets[0]
    assert first_bucket["key"] == index_names[1]
    assert first_bucket["doc_count"] == 1000

    second_bucket = buckets[1]
    assert second_bucket["key"] == index_names[0]
    assert second_bucket["doc_count"] == 1


@pytest.mark.asyncio
async def index_aggregate_group_by_term_test(client: Elasticsearch):
    """
    Terms Aggregation - Counting Unique Values.
    Scenario: You want to count the number of documents for each unique value in a specific field,
    such as analyzing the distribution of products in an e-commerce dataset.
    """
    query_size = 0  # The size of the return items.
    aggs = {"group_by_term": {"terms": {"field": "hasFlag"}}}

    response = await client.search(index=target_index, aggs=aggs, size=query_size)

    aggregations = response.body["aggregations"]
    assert aggregations is not None

    indexes = aggregations["group_by_term"]
    buckets = indexes["buckets"]
    assert len(buckets) == 2
    assert buckets[0]["doc_count"] == 666
    assert buckets[1]["doc_count"] == 334


@pytest.mark.asyncio
async def index_aggregate_date_histogram_test(client: Elasticsearch):
    """
    Date Histogram Aggregation - Time Series Analysis.
    Scenario: You have timestamped data, and you want to analyze it over time in intervals (e.g., daily, weekly, or monthly),
    such as tracking the number of orders over time.
    """
    query_size = 0  # The size of the return items.
    aggs = {
        "items_over_time": {
            "date_histogram": {
                "field": "timestamp",
                "calendar_interval": "month",
                "format": "yyyy-MM",
                "min_doc_count": 0,
            }
        }
    }

    response = await client.search(index=target_index, aggs=aggs, size=query_size)
    aggregations = response.body["aggregations"]

    items = aggregations["items_over_time"]
    buckets = items["buckets"]
    assert len(buckets) > 0


@pytest.mark.asyncio
async def index_aggregate_range_test(client: Elasticsearch):
    """
    Range Aggregation - Grouping by Numeric Ranges.
    Scenario: You want to group documents into numeric value ranges to analyze data distribution, such as grouping salaries into income brackets.
    """
    query_size = 0  # The size of the return items.
    aggs = {
        "seq_ranges": {
            "range": {
                "field": "seq",
                "ranges": [
                    {"from": 0, "to": 200},
                    {"from": 200, "to": 600},
                    {"from": 600},
                ],
            }
        }
    }

    response = await client.search(index=target_index, aggs=aggs, size=query_size)
    aggregations = response.body["aggregations"]

    items = aggregations["seq_ranges"]
    buckets = items["buckets"]
    assert len(buckets) > 0

File: node/src/__tests__/index_aggregates.test.ts
import { getClient } from "../elasticsearch_client";

const TARGET_INDEX = "test_index_2";

describe("Elastic search aggregates.", () => {
  const client = getClient();

  it("Count document count in multiple indexes", async () => {
    const indexNames = ["test_index_1", "test_index_2"];
    const querySize = 0; // Set the result set size to 0 to just get the aggregation results.
    const response = await client.search({
      index: indexNames.join(","),
      aggs: {
        indexes: {
          terms: {
            field: "_index",
            size: indexNames.length,
          },
        },
      },
      size: querySize,
    });

    const { aggregations } = response;
    const { indexes } = aggregations;
    const buckets = indexes["buckets"];
    expect(buckets.length === 2).toBeTruthy();

    const [bucket1, bucket2] = buckets;
    expect(bucket1["doc_count"] > 0).toBeTruthy();
    expect(bucket2["doc_count"] > 0).toBeTruthy();
  });

  it("Aggregate group by term", async () => {
    /**
      Terms Aggregation - Counting Unique Values.
      Scenario: You want to count the number of documents for each unique value in a specific field,
      such as analyzing the distribution of products in an e-commerce dataset.
     */
    const querySize = 0;
    const response = await client.search({
      index: TARGET_INDEX,
      aggs: {
        group_by_term: {
          terms: {
            field: "hasFlag",
          },
        },
      },
      size: querySize,
    });

    const { aggregations } = response;
    const items = aggregations["group_by_term"];
    const buckets = items["buckets"];
    expect(buckets.length === 2).toBeTruthy();

    const [bucket1, bucket2] = buckets;
    expect(bucket1["doc_count"] > 0).toBeTruthy();
    expect(bucket2["doc_count"] > 0).toBeTruthy();
  });

  it("Aggregate date histogram", async () => {
    /**
      Date Histogram Aggregation - Time Series Analysis.
      Scenario: You have timestamped data, and you want to analyze it over time in intervals (e.g., daily, weekly, or monthly),
      such as tracking the number of orders over time.
     */
    const querySize = 0;
    const response = await client.search({
      index: TARGET_INDEX,
      aggs: {
        items_over_time: {
          date_histogram: {
            field: "timestamp",
            calendar_interval: "month",
            format: "yyyy-MM",
            min_doc_count: 0,
          },
        },
      },
      size: querySize,
    });

    const { aggregations } = response;
    const items = aggregations["items_over_time"];
    const buckets = items["buckets"];
    expect(buckets.length > 0).toBeTruthy();
  });

  it("Aggregate range", async () => {
    /**
      Range Aggregation - Grouping by Numeric Ranges.
      Scenario: You want to group documents into numeric value ranges to analyze data distribution, such as grouping salaries into income brackets.
     */
    const querySize = 0;
    const response = await client.search({
      index: TARGET_INDEX,
      aggs: {
        seq_ranges: {
          range: {
            field: "seq",
            ranges: [{ from: 0, to: 200 }, { from: 200, to: 600 }, { from: 600 }],
          },
        },
      },
      size: querySize,
    });

    const { aggregations } = response;
    const items = aggregations["seq_ranges"];
    const buckets = items["buckets"];
    expect(buckets.length > 0).toBeTruthy();
  });
});
